File: kotlin/everybody_codes/src/main/kotlin/org/gristle/pdxpuzzles/everybodycodes/y2025/Y25D12.kt
package org.gristle.pdxpuzzles.everybodycodes.y2025

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.async
import kotlinx.coroutines.awaitAll
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.runBlocking
import org.gristle.pdxpuzzles.everybodycodes.utilities.Day
import org.gristle.pdxpuzzles.utilities.objects.StringGrid

object Y25D12 : Day {
    private fun parse(input: String): Pair<StringGrid, BooleanArray> {
        val clearing = StringGrid(input)
        val visited = BooleanArray(clearing.string.length)
        return clearing to visited
    }

    private fun solve(clearing: StringGrid, visited: BooleanArray, todo: MutableList<Int>): Int {
        var todo = todo
        var next = mutableListOf<Int>()
        for (pos in todo) {
            visited[pos] = true
        }

        while (todo.isNotEmpty()) {
            for (pos in todo) {
                val barrel = clearing.string[pos]
                for (adjPos in clearing.getNeighborIndices(pos)) {
                    if (!visited[adjPos] && barrel >= clearing.string[adjPos]) {
                        visited[adjPos] = true
                        next.add(adjPos)
                    }
                }
            }
            todo.clear()
            todo = next.also { next = todo }
        }

        return visited.count { it }
    }

    override fun part1(input: String): Int {
        val (clearing, visited) = parse(input)
        val start = mutableListOf(0)
        return solve(clearing, visited, start)
    }

    override fun part2(input: String): Int {
        val (clearing, visited) = parse(input)
        val start = mutableListOf(0, input.length - 1)
        return solve(clearing, visited, start)
    }
    override fun part3(input: String): Int = runBlocking(Dispatchers.Default) {
        val (clearing, visited) = parse(input)
        var history = visited.clone()
        val winners = mutableListOf<Int>()

        val peaks = clearing.string.withIndex()
            .filter { (idx, c) ->
                clearing.getNeighborIndices(idx).all { adj -> c > clearing[adj] }
            }.map { it.index }

        repeat(3) {
            val (_, exploded, winner) = coroutineScope {
                peaks
                    .filter { it !in winners }
                    .map { pos ->
                        async {
                            val exploded = history.clone()
                            val barrels = solve(clearing, exploded, mutableListOf(pos))
                            Triple(barrels, exploded, pos)
                        }
                    }.awaitAll()
                    .maxBy { (barrels) -> barrels }
            }
            winners.add(winner)
            history = exploded
        }
        solve(clearing, visited, winners)
    }
}

fun main() = Day.runDay(Y25D12::class)

//    Quest 1: 240 (2ms)
//    Quest 2: 5731 (7ms)
//    Quest 3: 4135 (210ms)
//    Total time: 220ms

File: rust/everybody-codes/src/bin/2025/y25d12.rs
use everybody_codes::utilities::inputs::get_event_inputs;
use itertools::Itertools;
use rayon::iter::{IntoParallelRefIterator, ParallelIterator};
use utilities::structs::{stopwatch::{ReportDuration, Stopwatch}, str_grid::{AdjacentMetadata, StrGrid}};

type Input<'a> = &'a str;
type Clearing<'a> = StrGrid<'a>;

fn main() {
    let mut stopwatch = Stopwatch::new();
    stopwatch.start();
    let (input1, input2, input3) = get_event_inputs(25, 12);
    println!("Input parsed ({})", stopwatch.lap().report());
    println!("1. {} ({})", part1(&input1), stopwatch.lap().report());
    println!("2. {} ({})", part2(&input2), stopwatch.lap().report());
    println!("3. {} ({})", part3(&input3), stopwatch.lap().report());
    println!("Total: {}", stopwatch.stop().report());
}

fn parse(input: Input<'_>) -> (Clearing<'_>, Vec<bool>) {
    let clearing: StrGrid<'_> = input.into();
    let visited = vec![false; clearing.s.len()];
    (clearing, visited)
}

fn solve(clearing: &StrGrid, visited: &mut Vec<bool>, mut todo: Vec<usize>) -> usize {
    let mut next = Vec::new();
    for &pos in todo.iter() {
        visited[pos] = true;
    }

    while !todo.is_empty() {
        for pos in todo.drain( .. ) {
            let barrel = clearing.s[pos];
            for AdjacentMetadata { pos: adj_pos, dir: _, b: adj_barrel } in clearing.adjacent(pos) {
                if !visited[adj_pos] && barrel >= adj_barrel {
                    visited[adj_pos] = true;
                    next.push(adj_pos);
                }
            }
        }
        std::mem::swap(&mut todo, &mut next);
    }
    visited.iter().filter(|&&it| it).count()
}

fn part1(input: Input) -> usize {
    let (clearing, mut visited) = parse(input);
    let start = vec![0usize];
    solve(&clearing, &mut visited, start)
}

fn part2(input: Input) -> usize {
    let (clearing, mut visited) = parse(input);
    let start = vec![0, input.len() - 1];
    solve(&clearing, &mut visited, start)
}

fn part3(input: Input) -> usize {
    let (clearing, mut visited) = parse(input);
    let mut history = visited.clone();
    let mut winners = Vec::new();

    let peaks = clearing.s.iter().enumerate()
        .filter(|&(idx, &c)| {
            clearing.adjacent(idx).all(|adj| c > adj.b)
        })
        .map(|(idx, _)| idx)
        .collect_vec();

    for _ in 0..3 {
        let (_, exploded, winner) = peaks
            .par_iter()
            .filter(|&pos| !winners.contains(pos))
            .map(|&pos| {
                let mut exploded = history.clone();
                let barrels = solve(&clearing, &mut exploded, vec![pos]);
                (barrels, exploded, pos)
            })
            .max_by_key(|(barrels, _, _)| *barrels)
            .unwrap();
        winners.push(winner);
        history = exploded;
    }
    solve(&clearing, &mut visited, winners)
}

#[test]
fn default() {
    let (input1, input2, input3) = get_event_inputs(25, 12);
    assert_eq!(240, part1(&input1));
    assert_eq!(5731, part2(&input2));
    assert_eq!(4135, part3(&input3));
}

// Input parsed (41μs)
// 1. 240 (30μs)
// 2. 5731 (92μs)
// 3. 4135 (8.774ms)
// Total: 8.943ms